### PR TITLE
fix: [CDS-80902]: add retries with exponential backoff for find file operation

### PR DESCRIPTION
**Pull request by meenaravichandran1**
## Describe your changes
Add retries with exponential backoff when findfile fails with null response, i.e, connectivity issues.

**Problem**
Informatica recently faced a git connectivity issue while trying to fetch file.

**Troubleshooting**
Inferring from the logs, in this specific scenario, the response received from the provider was null and the following is logged in the delegate https://cloudlogging.app.goo.gl/uT5QUPk2BGv16F1F6, executed block of code https://github.com/harness/harness-core/blob/master/product/ci/scm/file/file.go#L51.

We are sure that the line https://github.com/harness/harness-core/blob/master/product/ci/scm/file/file.go#L62 didn't get executed because for all statuses >=300, we print it in the delegate logs, code https://github.com/harness/harness-core/blob/release/delegate/805xx/952-scm-java-client/src/main/java/io/harness/impl/scm/ScmServiceClientImpl.java#L1156.

During the time period when this connectivity error happened, there are no error logs in the delegate for the same, reference https://cloudlogging.app.goo.gl/1LGYdGDZsYFmzfY17. Hence, sure that the response was null.

**Fix**
It is good to have retries for network-related failures for a few times, instead of failing the step or the pipeline. This adds exponential backoff to network related failures/when the response is null.

With the current data points, this is only happening for github and it gives the response as null, if we see some other failures, we can add it to the retry logic. 

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/54216)

### CODE REVIEW

**Comment by meenaravichandran1** [comment]
trigger uts

**Comment by meenaravichandran1** [comment]
trigger uts

File: product/ci/scm/file/file.go [hunk 1]
```diff
@@ -28,6 +28,11 @@ import (
 	"go.uber.org/zap"
 )
 
+const (
+	serverErrorRetryCount = 3
+	serverErrorRetrySleep = 3 * time.Second
```
Comment (rutvijmehta-harness) [inline product/ci/scm/file/file.go:33]: Do you need 3 second sleep? 100ms in git.go should be enough.

You can use these variables from git.go as well.

**Review by rutvijmehta-harness — DISMISSED** [review]
LGTM

**Review by rutvijmehta-harness — APPROVED** [review]

**Comment by shivagowda** [comment]
trigger review

**Review by shivagowda — APPROVED** [review]
lgtm

**Comment by shivagowda** [comment]
trigger review

**Pull request merged**